core/src/main/java/org/apache/iceberg/AllDataFilesTable.java

```diff
@@ -0,0 +1,149 @@
+ * A {@link Table} implementation that exposes a table's valid data files as rows.
+ * <p>
+ * A valid data file is one that is readable from any snapshot currently tracked by the table.
+ * <p>
+ * This table may return duplicate rows.
+ */
+public class AllDataFilesTable extends BaseMetadataTable {
+  private final TableOperations ops;
+  private final Table table;
+
+  public AllDataFilesTable(TableOperations ops, Table table) {
+    this.ops = ops;
+    this.table = table;
+  }
+
+  @Override
+  Table table() {
+    return table;
+  }
+
+  @Override
+  String metadataTableName() {
+    return "all_data_files";
+  }
+
+  @Override
+  public TableScan newScan() {
+    return new AllDataFilesTableScan(ops, table, schema());
+  }
+
+  @Override
+  public Schema schema() {
+    Schema schema = new Schema(DataFile.getType(table.spec().partitionType()).fields());
+    if (table.spec().fields().size() < 1) {
+      // avoid returning an empty struct, which is not always supported. instead, drop the partition field (id 102)
+      return TypeUtil.selectNot(schema, Sets.newHashSet(102));
+    } else {
+      return schema;
+    }
+  }
+
+  @Override
+  public String location() {
+    return table.currentSnapshot().manifestListLocation();
+  }
+
+  public static class AllDataFilesTableScan extends BaseTableScan {
+    private static final long TARGET_SPLIT_SIZE = 32 * 1024 * 1024; // 32 MB
+    private final Schema fileSchema;
+
+    AllDataFilesTableScan(TableOperations ops, Table table, Schema fileSchema) {
+      super(ops, table, fileSchema);
+      this.fileSchema = fileSchema;
+    }
+
+    private AllDataFilesTableScan(
+        TableOperations ops, Table table, Long snapshotId, Schema schema, Expression rowFilter,
+        boolean caseSensitive, boolean colStats, Collection<String> selectedColumns, Schema fileSchema,
+        ImmutableMap<String, String> options) {
+      super(ops, table, snapshotId, schema, rowFilter, caseSensitive, colStats, selectedColumns, options);
+      this.fileSchema = fileSchema;
+    }
+
+    @Override
+    protected TableScan newRefinedScan(
+        TableOperations ops, Table table, Long snapshotId, Schema schema, Expression rowFilter,
+        boolean caseSensitive, boolean colStats, Collection<String> selectedColumns,
+        ImmutableMap<String, String> options) {
+      return new AllDataFilesTableScan(
+          ops, table, snapshotId, schema, rowFilter, caseSensitive, colStats, selectedColumns, fileSchema, options);
+    }
+
+    @Override
+    public TableScan useSnapshot(long scanSnapshotId) {
+      throw new UnsupportedOperationException("Cannot select snapshot: all_data_files is for all snapshots");
+    }
+
+    @Override
+    public TableScan asOfTime(long timestampMillis) {
+      throw new UnsupportedOperationException("Cannot select snapshot: all_data_files is for all snapshots");
+    }
+
+    @Override
+    protected long targetSplitSize(TableOperations ops) {
+      return TARGET_SPLIT_SIZE;
```

Comment:
Seems like something we should map to a table/read property \w default as opposed to hard coded.

Reply:
I think we can add this later if necessary. I'd like to keep these simple and add features as we go.

Reply:
I believe it generally makes sense to configure the split size for metadata tables. In our tables, it is not uncommon to see 4-6 MB manifest files. If we have a reasonable cluster and allow split sizes to be 16 MB, metadata queries can be 2 times faster.

I've created #817 so that we can discuss it.